pallene/cfg.lua

```diff
@@ -0,0 +1,276 @@
+-- Copyright (c) 2020, The Pallene Developers
+-- Pallene is licensed under the MIT license.
+-- Please refer to the LICENSE and AUTHORS files for details
+-- SPDX-License-Identifier: MIT
+
+local Set = require "pallene.set"
+
+local CFG = {}
+--[[
+    A control flow graph is a representation of all paths 
+    that might be traversed through a program execution. 
+    
+    The nodes represent commands that use or alter data 
+    and the edges represent control instructions.
+
+    In this module, each node is defined by two sets of node indices (from and to)
+    and a list of commands. These commands are pointers to nodes in the ir module.
```

Comment:
I can change it for this model. In fact it is way cleaner. In my first implementation it was quite similar to that. But somewhere in the middle the OO bug bit me.

Something that I have been thinking lately is to use the ir.iter function. Now that I truly understand what I am doing it seems obvious :P

> Additionally, what do you think about making the source and sink nodes part of the list of nodes?

I would have to ignore them in the uses of the CFG and I don't really see benefits of having them on the list.

> What do you think about having the neighbor lists be just regular lists instead of sets?

Because they are sets (and in some situations, the same neighbor is added multiple times, such as in return). Now the sets are implemented as lists with duplicate removal (its terrible), but when I change for key sets, I think that it will be ok. In this commit all the uses are already in function format so I can change all by changing the functions in `Set.lua`

Reply:
> I would have to ignore them in the uses of the CFG and I don't really see benefits of having them on the list.

My motivation for asking that question is that having the node ID be either a number or a string (`"entry"`/`"exit"`) is a but funky. That said, I'll concede that ignoring the extra nodes is a bit annoying because Lua lacks a `continue` statement.

>  the same neighbor is added multiple times, such as in return). Now the sets are implemented as lists with duplicate removal (its terrible), but when I change for key sets, I think that it will be ok.

I think that for neighbor lists it is useful if the iteration order is deterministic (ipairs instead of pairs). I understand that we don't want to add an edge twice but we don't necessarily need to use the same Set datatype that we are using for our dataflow optimizations.

Since the neighbor lists are short an O(N) function would be OK. For example:

```
-- Add an element to a sorted list, if it is not in the list.
-- CFG nodes have low degree so O(N) insertion is OK.
function add_neighbor(xs, v)
    local N = #xs
    local i = 1
    while i <= N do
        local x = xs[i]
        if x == v then return end
        if x >  v then break  end
        i = i + 1
    end
    table.insert(xs, i, v)
end
```